src/Infrastructure.VS/ThreadHandling.cs

```diff
@@ -41,19 +40,26 @@ public class ThreadHandling : IThreadHandling
 
         public async Task RunOnUIThread(Action op) => await VS.RunOnUIThread.RunAsync(op);
 
-        public async Task<T> RunOnBackgroundThread<T>(Func<T> asyncMethod)
+        public async Task<T> RunOnBackgroundThread<T>(Func<Task<T>> asyncMethod)
         {
-            if (!ThreadHelper.CheckAccess())
-            {
-                return asyncMethod();
-            }
+            var startedOnUIThread = ThreadHelper.CheckAccess();
+
+            return await ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    if (startedOnUIThread)
+                    {
+                        await SwitchToBackgroundThread();
+                    }
+
+                    var result = await asyncMethod();
 
-            await SwitchToBackgroundThread();
-            var result = asyncMethod();
-            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    if (startedOnUIThread)
+                    {
+                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    }
 
-            Debug.Assert(ThreadHelper.CheckAccess(), "Expecting to returning on the UI thread");
-            return result;
+                    return result;
```

Comment:
This method is now compliant with [VSTHRD003](https://github.com/Microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD003.md). It can't accept a Task directly as that violates the rule.

Instead, it has to accept a factory method that returns the task, which can then be passed to `ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync`.

Note: there are currently no tests for this method.